lib/tpm2_auth_util.c

```diff
@@ -48,8 +50,91 @@ static bool handle_hex_password(const char *password, TPM2B_AUTH *auth) {
     return true;
 }
 
+static bool handle_password(const char *password, TPM2B_AUTH *auth);
+
+static tool_rc get_auth_for_file_param(const char* password, TPM2B_AUTH *auth) {
+    const char* path = password;
+    size_t size = (sizeof(auth->buffer) * 2) + HEX_PREFIX_LEN + 1;
+    bool is_a_tty = isatty(STDIN_FILENO);
+    UINT8 *buffer = calloc(1, size);
+
+    if (!buffer) {
+        LOG_ERR("oom");
+        return tool_rc_general_error;
+    }
+
+    if (path) {
+        path += FILE_PREFIX_LEN;
+        path = strcmp("-", path) ? path : NULL;
+    }
+    if (!is_a_tty || path) {
+        UINT16 fsize = sizeof(buffer) - 1;
+        bool ret = files_load_bytes_from_buffer_or_file_or_stdin(NULL, path,
+                                                                 &fsize, buffer);
+        if (!ret) {
+            free(buffer);
+            return tool_rc_general_error;
+        }
+        /* bash here strings and many commands add a trailing newline, if its stdin, kill the newline */
+        if (!path && buffer[fsize - 1] == '\n') {
+            buffer[fsize - 1] = '\0';
+        }
+    } else {
+        /*
+         * It is a TTY and we're reading from stdin.
+         * Prompt the user for the password with echoing
+         * disabled.
+         */
+        tool_rc rc = console_display_echo_control(false);
```

Comment:
Do we need a signal handler to restore echo-on mode on interrupt?

Reply:
@nicowilliams Thank you for the review, and apologies for the first sloppy version of the PR ;-)
There is already a signal handler for the FAPI commands. Next week I will create a separate PR for the tool commands if there is no handler.

Reply:
@nicowilliams I have changed the buffer allocation for getline from stack allocated to heap allocated to avoid reallocation errors from getline.
But I think we do not need an additional signal handler, because the program
is always terminated by SIGTERM or SIGINT and so all resources will be reclaimed.

Reply:
You need a signal handler to restore echo-on, otherwise if you ^C while being prompted then you'll get your shell prompt back with echo-off.

Reply:
Yes that's true. I had tested ^C in the bash but in sh there is the problem with echo-off.
I have added the solution from the fapi tool commands.